client/src/sim.rs

```diff
@@ -243,33 +280,50 @@ impl Sim {
     fn send_input(&mut self) {
         let velocity = sanitize_motion_input(self.orientation * self.average_velocity);
         let params = self.params.as_ref().unwrap();
-        let generation = self.prediction.push(
-            &(velocity
-                * params.movement_speed
-                * self.params.as_ref().unwrap().step_interval.as_secs_f32()),
-        );
+        let player_input = CharacterInput {
+            movement: velocity,
+            no_clip: self.no_clip,
+        };
+        let graph = &self.graph;
```

Comment:
Is this borrow splitting needed with modern Rust?

Reply:
If I unsplit this, I get the following error:
```
error[E0502]: cannot borrow `self.prediction` as mutable because it is also borrowed as immutable
   --> client\src\sim.rs:287:26
    |
287 |           let generation = self.prediction.push(&player_input, |position, input| {
    |                            ^               ----                ----------------- immutable borrow occurs here
    |                            |               |
    |  __________________________|               immutable borrow later used by call
    | |
288 | |             CharacterControllerPass {
289 | |                 position,
290 | |                 input,
291 | |                 graph: &self.graph,
    | |                         ---------- first borrow occurs due to use of `self` in closure
...   |
295 | |             .step()
296 | |         });
    | |__________^ mutable borrow occurs here
```

I'm not yet sure why this is the case, but my guess is that while modern rust has gotten good about splitting fields, `self` might still be borrowed as a single unit in the implicit closure struct.

Reply:
What version of rustc are you using?

Reply:
I'm using 1.66.0. I'm certain my IDE is also using it because it detected the new clippy lints. (I didn't notice the clippy lints before creating this PR because I was still on 1.65 then, but I can still reproduce this error.)

Reply:
Weird. Well, no matter.

Reply:
Oh, I looked further into this. If we switch to the 2021 edition of Rust, I believe the error goes away. Currently, we're using the 2018 edition. I assume that means that this smarter capture logic is not backwards compatible.

Reply:
Ahah, nice catch! Edition updates are usually pretty painless, but no rush.